src/librustc_trans/back/linker.rs

```diff
@@ -164,27 +197,26 @@ impl<'a> Linker for GnuLinker<'a> {
         self.hint_static();
         let target = &self.sess.target.target;
         if !target.options.is_like_osx {
-            self.cmd.arg("-Wl,--whole-archive")
-                    .arg("-l").arg(lib)
-                    .arg("-Wl,--no-whole-archive");
+            self.linker_arg("--whole-archive").cmd.arg("-l").arg(lib);
```

Comment:
FYI: this style (e.g. `"-l" "System"`) isn't supported by ld64 on darwin - that linker requires the style `"-lSystem"`. I'm testing a PR that changes this.